### PR TITLE
[SPARK-14565][ML] RandomForest should use parseInt and parseDouble for feature subset size instead of regexes

### PR DESCRIPTION
**Pull request by yongtang**
## What changes were proposed in this pull request?

This fix tries to change RandomForest's supported strategies from using regexes to using parseInt and
parseDouble, for the purpose of robustness and maintainability.
## How was this patch tested?

Existing tests passed.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 1]
```diff
@@ -397,8 +399,31 @@ private[spark] object RandomForestParams {
   final val supportedFeatureSubsetStrategies: Array[String] =
     Array("auto", "all", "onethird", "sqrt", "log2").map(_.toLowerCase)
 
-  // The regex to capture "(0.0-1.0]", and "n" for integer 0 < n <= (number of features)
-  final val supportedFeatureSubsetStrategiesRegex = "^(?:[1-9]\\d*|0?\\.\\d*[1-9]\\d*|1\\.0+)$"
+  object integerFeatureSubsetStrategy {
+    def unapply(strategy: String): Option[Int] = try {
+      val number = strategy.toInt
+      if (0 < number) {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Nit: number > 0? and import java.lang.NumberFormatException

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/DecisionTreeMetadata.scala [hunk 2]
```diff
@@ -184,15 +185,15 @@ private[spark] object DecisionTreeMetadata extends Logging {
...
-    val isIntRegex = "^([1-9]\\d*)$".r
-    val isFractionRegex = "^(0?\\.\\d*[1-9]\\d*|1\\.0+)$".r
     val numFeaturesPerNode: Int = _featureSubsetStrategy match {
       case "all" => numFeatures
       case "sqrt" => math.sqrt(numFeatures).ceil.toInt
       case "log2" => math.max(1, (math.log(numFeatures) / math.log(2)).ceil.toInt)
       case "onethird" => (numFeatures / 3.0).ceil.toInt
-      case isIntRegex(number) => if (BigInt(number) > numFeatures) numFeatures else number.toInt
-      case isFractionRegex(fraction) => (fraction.toDouble * numFeatures).ceil.toInt
+      case RandomForestParams.integerFeatureSubsetStrategy(number) =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/DecisionTreeMetadata.scala]: I think we should simplify the checks.

~~~scala
import scala.util.Try

val numFeaturesPerNode: Int = _featureSubsetStrategy match {
   ...
   case _ =>
     Try(_featureSubsetStrategy.toDouble).toOption match {
       case Some(value) =>
         if (value <= 0.0) {
           throw new IllegalArgumentException(...)
         } else if (value < 1.0) {
           math.ceil(value * numFeatures)
         } else {
           math.ceil(value)
         }
       case None =>
         throw new IllegalArgumentException(...)
     }
}
~~~

See my comment below for the validator.

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 3]
```diff
@@ -346,10 +346,12 @@ private[ml] trait HasFeatureSubsetStrategy extends Params {
    */
   final val featureSubsetStrategy: Param[String] = new Param[String](this, "featureSubsetStrategy",
     "The number of features to consider for splits at each tree node." +
-      s" Supported options: ${RandomForestParams.supportedFeatureSubsetStrategies.mkString(", ")}",
+      s" Supported options: ${RandomForestParams.supportedFeatureSubsetStrategies.mkString(", ")}" +
+      s", (0.0-1.0], [1-n].",
     (value: String) =>
       RandomForestParams.supportedFeatureSubsetStrategies.contains(value.toLowerCase)
-      || value.matches(RandomForestParams.supportedFeatureSubsetStrategiesRegex))
+      || RandomForestParams.integerFeatureSubsetStrategy.unapply(value).isDefined
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: `Try(value.toDouble).filter(_ > 0.0).isSuccess`

**Comment by mengxr** [comment]
ok to test

**Comment by mengxr** [comment]
Besides inline comments, we should also consider the behavior at `1.0`. Since `all` is already an option, maybe we should treat `1.0` as `numFeaturesPerNode = 1` instead of `all`.

**Comment by jkbradley** [comment]
@mengxr I'd prefer to treat "1" as numFeaturesPerNode = 1 and "1.0" as equivalent to "all" in order to match sklearn's semantics.

**Comment by jkbradley** [comment]
One more comment: I'd like to make sure there is always at least 1 feature being used.  Could you please update the parsing and add that to the unit test in RandomForestSuite.scala?  Thanks!

**Comment by yongtang** [comment]
Thanks @srowen @mengxr @jkbradley. The pull request has been updated. Please let me know if there are any further issues.

**Comment by mengxr** [comment]
LGTM pending Jenkins

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**